### PR TITLE
Generate TextX compatible Classes

### PR DESCRIPTION
**Pull request by fgro93**
Changes:

- Add a cli option `--textX` to trigger textX compatible code generation.

- Override  `EEnum`, `EDataType`, and `EObject` to keep compatibility with textX.

- Ecore enums result in own classes.

- Disable kwargs check, because textX will always pass a parent parameter

- Disable None check for parameters

- Add `classes` variable

### CODE REVIEW

**Comment by moltob** [comment]
@aranega I leave it up to you to decide about this PR from a functional POV. If you want to accept it, we should split the review. I can make sure it fits the current code well, but I have no experience with textX yet. So you should look at the "what".

Please let me know.

**Comment by aranega** [comment]
@moltob No problem, I will do this during lunch today ;)

File: pyecoregen/templates/module_utilities.tpl [hunk 1]
```diff
@@ -77,8 +88,12 @@ class Derived{{ d.name | capitalize }}(EDerivedCollection):
 
 {%- macro generate_feature_init(feature) %}
     {%- if feature.upperBound == 1 %}
+        {%- if not textX %}
         if {{ feature.name }} is not None:
             self.{{ feature.name }} = {{ feature.name }}
+        {%- else %}
```
Comment (aranega) [inline pyecoregen/templates/module_utilities.tpl]: This `if` statement removal breaks the default value capabilities of `EStructuralFeatures`, is there a reason why the affectation is mandatory? Perhaps textX directly works with instance `__dict__` sometimes?

File: pyecoregen/templates/module.py.tpl [hunk 2]
```diff
@@ -20,12 +20,49 @@ eClassifiers = {}
...
+                     not a.startswith('__') and not
+                     a.startswith('_tx_')]:
+            obj = getattr(self, attr)
+            if obj and isinstance(obj, str):
+                return obj
+{% endif %}
 {{ modutil.generate_enum(c) }}
 {%- endfor %}
 {% for c in element.eClassifiers if c is type(ecore.EDataType) -%}
+{% if loop.first and textX %}
```
Comment (aranega) [inline pyecoregen/templates/module.py.tpl]: Wouldn't it be easier to simply generates a `MyDataTypeName.__name__ = 'MyDataTypeName'` for each `EDataType` generation?

File: pyecoregen/templates/module.py.tpl [hunk 3]
```diff
@@ -20,12 +20,49 @@ eClassifiers = {}
...
+        self = Ecore.EDataType.__new__(cls, *args, **kwargs)
+        self.__name__ = args[0]
+        return self
+
+{% endif %}
 {{ modutil.generate_edatatype(c) }}
 {%- endfor %}
 
 {%- for c in classes -%}
+{% if loop.first and textX %}
```
Comment (aranega) [inline pyecoregen/templates/module.py.tpl]: If I understood correctly (and I remember from textX), the idea is to put all unknown `kwargs` as simple instance attribute?
Anyway, I would propose:

```python
def __init__(self, **kwargs):
    super().__init__(**kwargs)
    for key, value in kwargs.items():
        if key not in self.__dict__:
            continue
        setattr(self, key, value)
```
What do you think?

File: pyecoregen/templates/module.py.tpl [hunk 4]
```diff
@@ -20,12 +20,49 @@ eClassifiers = {}
 getEClassifier = partial(Ecore.getEClassifier, searchspace=eClassifiers)
 
 {%- for c in element.eClassifiers if c is type(ecore.EEnum) -%}
+{% if loop.first and textX %}
```
Comment (aranega) [inline pyecoregen/templates/module.py.tpl]: I'm not sure I get the point with the `EEnum`. From the quick experiment I made, I didn't get the usage for the `EEnum` redefinition and the `EEnum` instance promotion to a class level. In my test, I just added `MY_ENUM.__name__ = 'MY_ENUM'` for each `EEnum` instance and it seemed to worked. Well, partially, but not because of your code or because of the modification, but because of a bug in PyEcore (I need to patch it quickly), which does not handle `string` attributes as it should when the feature is typed by an `EEnum`.
Should the enumeration be promoted to a class level?

File: pyecoregen/templates/module_utilities.tpl [hunk 5]
```diff
@@ -1,5 +1,16 @@
 {%- macro generate_enum(e) %}
-{{ e.name }} = EEnum('{{ e.name }}', literals=[{{ e.eLiterals | map(attribute='name') | map('pyquotesingle') | join(', ') }}])
+{% set name = e.name %}
+{% if textX %}
```
Comment (aranega) [inline pyecoregen/templates/module_utilities.tpl]: Same as before, I'm not very sure I got it about the `EEnum`, but perhaps I missed something. If you have a small example, it would be great (currently, I only tested on the `library.ecore` metamodel).

File: pyecoregen/templates/module.py.tpl (hunk 2 above)
Comment (fgro93) [reply pyecoregen/templates/module.py.tpl]: Yes, would be way easier. I will change that.

File: pyecoregen/templates/module_utilities.tpl (hunk 1 above)
Comment (fgro93) [reply pyecoregen/templates/module_utilities.tpl]: My custom scope providers break, because instead of `None` I sometimes have a `property` object.

Consider this grammar:
```
TypeReference:
	predefined=BasicType | derived=[Type];

BasicType:
	Boolean = /Boolean(?=\W)/ |
	String = /String(?=\W)/ |
	Integer = /Integer(?=\W)/ ;

Type: StructType | TypeDefinition;
```

A resolved TypeReference with a `derived` reference looks like this:
```
TypeReference: {
  predefined: <property object>
  derived: <TypeDefinition object>
}
```

Is this the intended behavior?

File: pyecoregen/templates/module.py.tpl (hunk 3 above)
Comment (fgro93) [reply pyecoregen/templates/module.py.tpl]: Yap, textX puts all unknown kwargs as instance attributes.

I change it, like the readability of your code.

File: pyecoregen/templates/module.py.tpl (hunk 4 above)
Comment (fgro93) [reply pyecoregen/templates/module.py.tpl]: textX does not know about enums. An "enum" in textX looks like:
```
BasicType:
	Boolean = /Boolean(?=\W)/ |
	String = /String(?=\W)/ |
	Integer = /Integer(?=\W)/ ;
```
A textX object with value `String` of this rule looks like this:
`
BasicType: {
Boolean: '',
String: 'String',
Integer: '',
}
`

I wanted to create a model that behaves as a textX and an ecore model. To preserve those properties, I used classes. The added `value` property is just syntactical sugar, can be deleted.

Comment (fgro93) [reply pyecoregen/templates/module.py.tpl]: Your simple `__name__` solution resulted in an error:
```
  File "..\textx\model.py", line 598, in parse_tree_to_objgraph
    model = process_node(parse_tree)
  File "..\textx\model.py", line 397, in process_node
    process_node(n)
  File "..\textx\model.py", line 492, in process_node
    value = process_node(n)
  File "..\textx\model.py", line 397, in process_node
    process_node(n)
  File "..\textx\model.py", line 492, in process_node
    value = process_node(n)
  File "..\textx\model.py", line 397, in process_node
    process_node(n)
  File "..\textx\model.py", line 472, in process_node
    value = process_node(node[0])
  File "..\textx\model.py", line 372, in process_node
    inst = user_class.__new__(user_class)
  File "..\pyecore\ecore.py", line 137, in __new__
    instance = super().__new__(cls)
TypeError: object.__new__(X): X is not a type object (EEnum)
```

pyecore 0.9.0

Comment (aranega) [reply pyecoregen/templates/module.py.tpl]: Ok, got it for the `EEnum`! Regarding the simple solution, strange, I must have changed other stuffs during the quick experiment, sorry about that, I'll check this. Don't worry about the syntactical sugar, it can stay, I wasn't sure to understand what was the point. I think that the code could perhaps be simplified a little bit, and there is a collateral effect that should be addressed ;). I quickly test and I tell you about it.

Comment (aranega) [reply pyecoregen/templates/module.py.tpl]: @fgro93 I was doing some experiments, but then, re-reading your snippet, it hit me, I'm not sure that 

```
BasicType:
	Boolean = /Boolean(?=\W)/ |
	String = /String(?=\W)/ |
	Integer = /Integer(?=\W)/ ;
```

should be considered an enumeration. This instead should be:

```
BasicType:
	"Boolean" | "String" | "Integer";
```

You case looks more like an `EDataType` mapped to an 'object' instance?

File: pyecoregen/templates/module_utilities.tpl (hunk 1 above)
Comment (fgro93) [reply pyecoregen/templates/module_utilities.tpl]: I cannot, but here you can see a small working example:
https://github.com/fgro93/pyecore-Provider

```
class ProblemRuleValProvider():
    def __call__(self, obj, attr, obj_ref):
        reference = obj.parent.reference
        if reference.predefined:
            # imagine meaningful code here
            obj.attribute = 42
        print(hasattr(obj, 'attribute')) # prints True
        return reference
```

File: pyecoregen/templates/module.py.tpl (hunk 4 above)
Comment (fgro93) [reply pyecoregen/templates/module.py.tpl]: True. This is just needed if I want to use a custom class for this rule.
```
BasicType:
	"Boolean" | "String" | "Integer";
```

Problem here: textX won't use our ecore class anymore because `BasicType` does not have any properties. Without properties, an object instantiation is not necessary, therefore custom classes can be ignored.

Result:
```
  File "..\pyecore\valuecontainer.py", line 83, in _set
    self.check(value)
  File "..\pyecore\valuecontainer.py", line 47, in check
    raise BadValueError(value, self.feature.eType)
pyecore.valuecontainer.BadValueError: ("Expected type <class 'mymodel.mymodel.BasicType'>, but got type str with value Boolean instead", 'for class TypeReference')
```

`pyecore.valuecontainer.BadValueError: ("Expected type <class 'mymodel.mymodel.BasicType'>, but got type str with value Boolean instead", 'for class TypeReference')`

Comment (fgro93) [reply pyecoregen/templates/module.py.tpl]: @aranega please see:
https://github.com/fgro93/pyecore-Provider/tree/enum
 (Attention use the `enum` branch!)

**Comment by moltob** [comment]
I'll let you guys sort out generation semantics first. Please ping me once you're done, so I'll take a look at the "how".  Once code is fine, unit tests have to be added as well.

File: pyecoregen/templates/module.py.tpl (hunk 4 above)
Comment (aranega) [reply pyecoregen/templates/module.py.tpl]: @fgro93 Awesome! Thanks for the repository and the examples. Ok, I think the exception you have here is related to the bug in PyEcore. I'm checking this! Also, sorry if I seem picky, I really like your solution of 're-shaping' the PyEcore objects so they match textX and PyEcore, so I really would like to see it dealing with all corner cases ;). Thanks again @fgro93 !

Comment (aranega) [reply pyecoregen/templates/module.py.tpl]: I managed to make it work! I mean... The PyEcore bug is still here, let's be honest, but at the end the model is created as it should. To do this, I modified this part of the generated code:

```python
class EEnum(Ecore.EEnum):

    def __init__(self, name=None, default_value=None, literals=None, **kwargs):
        super().__init__(name, default_value, literals)
        self.__dict__.update(kwargs)


BasicTypeENUM = EEnum('BasicType', literals=['Boolean', 'String', 'Integer'])


class BasicType(EEnum):

    def __init__(self, name=None, default_value=None, literals=None, **kwargs):
        super().__init__(**kwargs)
        self.__dict__.update(BasicTypeENUM.from_string(self.value).__dict__)
```

to this:

```
BasicType = EEnum('BasicType', literals=['Boolean', 'String', 'Integer'])
BasicType.__name__ = 'BasicType'
```

And it seemed to be enough. There is perhaps some corner cases however where this fails, it is clear that tests would be required.

**Comment by moltob** [comment]
Having thought about this a bit more, I am not sure we should tie this extension this close to pyecoregen.

Let's assume @aranega and @fgro93 come to an agreement how to best weave in the textX dependencies. Then, tests would need to be written that assert the generated code (unfortunately, the current test gap is not showing up on coveralls at all, since it is not measuring coverage on generator templates).

These tests might depend on textX, to make sure a corresponding grammar works as expected. But that new dependency is heavy and I would not recommend to have it. Alternatively tests would simply assert a certain feature (like the enumeration base class) has been correctly generated. In that case however, we are mocking textX requirements in free-floating way. Meaning, the tests would still be green even if textX at some point has different requirements or we assume the wrong requirements in the first place. In short, we cannot really make sure textX support is working as expected without the dependency...

So we should possibly consider moving the extension to another package/library. What are the alternatives for such an approach? I see these:

* You can derive a new class from `EcoreGenerator`. For that class you can add new arguments as needed. You can also use updated templates as you see fit. If you want to, you can even call into the original templates like `module.tpl` does with `module_utilities.tpl` and only change what is really needed.
* I've seen that at least some of the changes in this PR could be implemented via mixins created by the `--user-module` option. Here you have the ability to inject alternative implementations for all generated classes. E.g. this should be able to deal with the enum base class.

What do you guys think?

**Comment by aranega** [comment]
Indeed, you are probably right, I was wondering the same thing about the direct integration in pyecoregen and the tests. It kinda make sense to have these modifications as an extensions, especially regarding dependencies. Could the extension be proposed as an optional dependency of pyecoregen?

After that, I don't know which one of the two approach you propose would be the best. The two implies advantages: the first solution seems more flexible (anticipating future textX modifications), while the second one would make a good use of existing mechanisms. What's your opinion @fgro93 ?

Regarding tests, a first test that is relevant would be to play the existing tests on the code generated with the `--textX` option enabled. This would show that the semantic is preserved between the adapted classes and the original ones.

File: pyecoregen/templates/module.py.tpl (hunk 4 above)
Comment (fgro93) [reply pyecoregen/templates/module.py.tpl]: @aranega Did you change something in pyecore to make this work? I cannot reproduce this as I still get the error: `TypeError: object.__new__(X): X is not a type object (EEnum)`.

I honestly value ecore behavior more than textX behavior, but this approach results in textX `incompatible` obejcts. Not sure if keeping the textX behvior is mandatory here, but I'd love to keep at least the `parent`-child relation.

**Comment by fgro93** [comment]
First thank you both for the input! 

I would prefer the first solution. I just feels more intuitive to me. I might also break the `--user-module` option if `--textX` is enabled, depending on the necessary template changes. If this is the case, users have to inherit from the generated classes and register them to xtext to work around this issue.

File: pyecoregen/templates/module_utilities.tpl (hunk 1 above)
Comment (fgro93) [reply pyecoregen/templates/module_utilities.tpl]: I see. Sounds like a good solution to me. :)

I agree, code changes by  `--textX` should be kept to a minimum.

File: pyecoregen/templates/module.py.tpl (hunk 4 above)
Comment (fgro93) [reply pyecoregen/templates/module.py.tpl]: Yes, that was my issue. I used the wrong grammar. My bad, sorry. It works now.

Comment (aranega) [reply pyecoregen/templates/module.py.tpl]: I'm relieved :), I was started to think that my machine was the "golden one" where nothing could fail. This make me think that perhaps, a kind of "validator" that could take the grammar, take the metamodel and points the incorrect matches between a rule and a class could be interesting.

**Comment by moltob** [comment]
Ok... in that case the extension should be added as part of another repository to keep pyecoregen completely independent of textX.

**Comment by fgro93** [comment]
@moltob @aranega 
I implemented a small repo:
https://github.com/fgro93/textx-pyecoregen

I won't be able to work on this anymore. I used as much code/templates as possible from pyecore.

if pyecoregen would provide a `createEcoreGenerator` the code could be reduced even further.
```
createEcoreGenerator(parsed_args)
    return EcoreGenerator(
        auto_register_package=parsed_args.auto_register_package,
        user_module=parsed_args.user_module,
        with_dependencies=parsed_args.with_dependencies
       ).generate(model, parsed_args.out_folder)
```

The PR can be closed now.

**Comment by aranega** [comment]
@fgro93 Thanks a lot for all your work! I will write a small documentation as soon as I find time for it, push it as PR. Also, I will try to write some tests.

@moltob Do you think that it could be a good idea to propose textx-pyecoregen as an extra dependency for pyecoregen?
